### PR TITLE
ENH: add optional argument `dtype` to numpy.fft.fftfreq (#9126)

### LINKED ISSUE #9126: ENH: Optional dtype for fftfreq/rfftfreq
Would be nice to be able to specify a `dtype` for `fftfreq`/`rfftfreq` other than `float64` (e.g. `float32`), to which it appears to default.

### PR DESCRIPTION
**Pull request by cval26**
Deals with issue #9126 requesting the addition of dtype argument option for `fftfreq` and `rfftfreq` functions. To do so, I added an optional argument and used the `reciprocal` function to set the type of the array that will return the result. 

A different way to do so would be to leave the calculation as is and instead modify the return line so that it reads `return asarray(results*val, dtype=dtype)`; I think the two ways should be equivalent in terms of the end result, but I was hoping for some feedback. 

I've submitted a PR draft since I'm not entirely sure how to do the check that the required `dtype` is one of the numpy float dtypes; I'm using the `isinstance` function to check, but I'm not sure what would be the proper types to check against and where to import those from. Any additional comments are of course welcome! 

EDIT: I'm now allowing the output array type to be either float32 or float64, with float64 being the default behavior (as before). Let me know what you think; I'm changing the PR from draft to open. 

cc @jakirkham @mhvk @eric-wieser @leofang @charris 

Closes #9126

### CODE REVIEW

**Comment by jakirkham** [comment]
Should we cast `d` to the same `dtype`?

**Comment by cvav-git** [comment]
We could do that, before using it to compute `val`, but I'm not sure if that would add something to the functionality; what's your thinking here?

**Comment by rossbar** [comment]
It's worth noting that even after the switch to pocketfft, `numpy.fft` still internally promotes `float32` -> `float64` (and analogously for `complex64`) as mentioned in https://github.com/numpy/numpy/issues/9126#issuecomment-302097508. That's not *necessarily* a blocker, but it is worth considering whether a `dtype=` kwarg should be supported for "helper functions" like `fftfreq` when the main functions in the fft module only support `float64` (`complex128`, etc).

I just wanted to comment here quickly, even if the mailing list might be the more appropriate forum for this discussion.

**Comment by jakirkham** [comment]
> We could do that, before using it to compute `val`, but I'm not sure if that would add something to the functionality; what's your thinking here?
> 
> **EDIT**: @jakirkham thoughts?

No strong thoughts. More of a question. Though you're right. This may not be needed.

**Comment by jakirkham** [comment]
LGTM. Thoughts from others?

File: numpy/fft/helper.py [hunk 1]
```diff
@@ -216,7 +227,11 @@ def rfftfreq(n, d=1.0):
     """
     if not isinstance(n, integer_types):
         raise ValueError("n should be an integer")
-    val = 1.0/(n*d)
+    if dtype != float32:
+        dtype = float64
+
+    val = reciprocal(n*d, dtype=dtype)
     N = n//2 + 1
     results = arange(0, N, dtype=int)
```
Comment (jakirkham) [inline numpy/fft/helper.py:235]: Sorry one last thought. Would it make sense to cast this to `dtype` as well?

```suggestion
    results = arange(0, N, dtype=dtype)
```

Comment (cvav-git) [reply numpy/fft/helper.py:235]: we could but from what I understand this is always just an array of integers, which when multiplied with the `dtype` cast `val` will lead to the result being of type `dtype` anyway, right?

**Comment by cvav-git** [comment]
> It's worth noting that even after the switch to pocketfft, `numpy.fft` still internally promotes `float32` -> `float64` (and analogously for `complex64`) as mentioned in [#9126 (comment)](https://github.com/numpy/numpy/issues/9126#issuecomment-302097508). That's not _necessarily_ a blocker, but it is worth considering whether a `dtype=` kwarg should be supported for "helper functions" like `fftfreq` when the main functions in the fft module only support `float64` (`complex128`, etc).
> 
> I just wanted to comment here quickly, even if the mailing list might be the more appropriate forum for this discussion.

that's a very good point @rossbar, thanks for bringing it up. To be honest I haven't really considered it this way; I saw that this was an open issue for quite some time, so I thought I would attempt to work on it. 

But considering the `fft` module as a whole, I agree that having helper functions offer more kwargs than the "main" ones is certainly odd. Since `fft` still always promotes types to the highest accuracy, I believe it's best if we also leave the helper functions in their current state; if at some point later on it's decided that a `dtype` kwarg should be offered throughout, then we can reopen an issue like that and apply such a change in all `fft` functions. Does that sound reasonable?

**Comment by hexane360** [comment]
It seems like we can revisit this now that numpy.fft supports half and single-precision (PR #25711). I would be interested in this mainly to have parity with other array API implementations that already support this (e.g. [jax](https://docs.jax.dev/en/latest/_autosummary/jax.numpy.fft.fftfreq.html)).
